store-pg/src/main/scala/com/socrata/pg/store/PGSecondary.scala

```diff
@@ -645,14 +657,81 @@ class PGSecondary(val config: Config) extends Secondary[SoQLType, SoQLValue] wit
 
   private def updateRollups(pgu: PGSecondaryUniverse[SoQLType, SoQLValue],
                             originalCopyInfo: Option[TruthCopyInfo],
-                            copyInfo: TruthCopyInfo) = {
+                            copyInfo: TruthCopyInfo,
+                            rollupDirty: RollupDirty) = {
+    if (rollupDirty.affectAll || originalCopyInfo.isEmpty) {
+      updateRollupsSlow(pgu, originalCopyInfo, copyInfo, rollupDirty)
+    } else {
+      updateRollupsFast(pgu, originalCopyInfo, copyInfo, rollupDirty),
+    }
+  }
+
+  private def updateRollupsSlow(pgu: PGSecondaryUniverse[SoQLType, SoQLValue],
+                                originalCopyInfo: Option[TruthCopyInfo],
+                                copyInfo: TruthCopyInfo,
+                                rollupDirty: RollupDirty): Unit = {
+
     val rm = new RollupManager(pgu, copyInfo)
     rm.dropRollups(immediate = true)
     pgu.datasetMapReader.rollups(copyInfo).foreach { ri =>
       rm.updateRollup(ri, originalCopyInfo, copyInfo.dataVersion)
     }
   }
 
+  /**
+   * TODO: Handle each changed rollup separately fallback each blocked one to copy table
+   *       without switching back to the slow path for the whole batch.
+   */
+  private def updateRollupsFast(pgu: PGSecondaryUniverse[SoQLType, SoQLValue],
+                                originalCopyInfo: Option[TruthCopyInfo],
+                                copyInfo: TruthCopyInfo,
+                                rollupDirty: RollupDirty): Unit = {
+
+    val rm = new RollupManager(pgu, copyInfo)
+    val rollups = pgu.datasetMapReader.rollups(copyInfo).foldLeft(Map.empty[String, com.socrata.datacoordinator.truth.metadata.RollupInfo]) { (acc, ri) =>
+      acc + (ri.name.underlying -> ri)
+    }
+
+    val unchanged = rollups -- rollupDirty.update.map(_.name) -- rollupDirty.delete.map(_.name)
+
+    val allCopyNumbersMatch = originalCopyInfo match {
+      case Some(oldCopyInfo) =>
+        unchanged.values.forall { ri =>
+          oldCopyInfo.copyNumber == ri.copyInfo.copyNumber
+        }
+      case None =>
+        false
+    }
+
+
+    if (!allCopyNumbersMatch) {
+      updateRollupsSlow(pgu, originalCopyInfo, copyInfo, rollupDirty);
+    } else {
+      rollupDirty.update.foreach { ri =>
+        rollups.get(ri.name) match {
+          case Some(r) =>
+            rm.dropRollup(r, true)
+            rm.updateRollup(r, originalCopyInfo, copyInfo.dataVersion)
+          case None =>
+            logger.warn(s"cannot find rollup ${ri.name} - ${copyInfo.datasetInfo.systemId}")
+        }
+      }
+
+      rollupDirty.delete.foreach { ri =>
```

Comment:
Is it possible to end up with both delete and update operations in one version?  In that case it seems we would need more logic to avoid removing rollups that where removed then updated (ie. created).  I'm not sure if the version combining code could result in that.

Reply:
good call...should just assume it can happen and handle it.